angel.py

```diff
@@ -75,131 +78,186 @@ def read_csv(filename):
-        interests = row[index_dict["interests"]].strip()
-        two_truths_one_lie = row[index_dict["two_truths_one_lie"]].strip()
-        introduction = row[index_dict["introduction"]].strip()
-        house_number = row[index_dict["house_number"]].strip().lower()
-        cg_number = row[index_dict["cg_number"]].strip().lower()
-        year_of_study = row[index_dict["year_of_study"]].strip().lower()
-        faculty = row[index_dict["faculty"]].strip().lower()
+        special_columns = column_names_index_dict["special columns"]
+        enum_columns = column_names_index_dict["enum columns"]
+        text_columns = column_names_index_dict["text columns"]
+
+        player_data = (generate_column_key_value_pair(special_columns, row) +
+                       generate_column_key_value_pair(text_columns, row) +
+                       generate_column_key_value_pair(enum_columns, row, check_enum_columns_input_valid))
+
+        print("output is\n", player_data)
+
     except FileNotFoundError as e:
         print("WARNING: Config file is named incorrectly or does not exist.")
         exit()
     except KeyError as e:
-        print("WARNING: Key in player_attribute_index of config.yml should not be changed. Key used here should match "
+        print("WARNING: Key in generalized_columns of config.yml should not be changed. Key used here should match "
               "those in the config.yml dictionary.")
         exit()
     except IndexError as e:
-        print("WARNING: Index value in player_attribute_index of config.yml is not within range of given playlist.csv."
+        print("WARNING: Index value in generalized_columns of config.yml is not within range of given playlist.csv."
               " Pls check if columns correspond to their respective indexes.")
         exit()
 
-    return Player(username=player_username,
-                  playername=player_name,
-                  genderpref=gender_pref,
-                  genderplayer=gender_player,
-                  interests=interests,
-                  twotruthsonelie=two_truths_one_lie,
-                  introduction=introduction,
-                  housenumber=house_number,
-                  cgnumber=cg_number,
-                  yearofstudy=year_of_study,
-                  faculty=faculty,)
+    return Player(player_data)
+
+
+def generate_column_key_value_pair(type_of_column, row, function=None):
+    output = []
+    for column_name in type_of_column:
+        # obtains the value of passing the column_name as a key to the type of column
+        column_dict = type_of_column[column_name]
+        if isinstance(column_dict, int):
+            # if it column_dit is an integer, it is because it is from special columns or text columns
+            row_input = row[column_dict]
+        else:
+            row_input = row[column_dict['index']]
+        if function is not None:
+            valid = function(type_of_column, column_dict, row_input.strip().lower())
```

Comment:
row_input.strip().lower() is called multiple times; it can be abstracted

Reply:
I don't quite see why you need to pass check_enum_columns_input_valid as a parameter. It seems to be the only case where this parameter is ever used?